spec.bs

```diff
@@ -403,6 +407,101 @@ Each {{AppHistory}} object has an associated <dfn for="AppHistory">navigate meth
+
+  1. If [=this=]'s [=AppHistory/entry list=] does not contain any {{AppHistoryEntry}} whose [=AppHistoryEntry/session history entry=]'s [=session history entry/app history key=] equals |key|, then return [=a promise rejected with=] an "{{InvalidStateError}}" {{DOMException}}.
+
+  1. Return the result of [=performing an app history traversal=] given [=this=], |key|, and |options|.
+</div>
+
+<div algorithm>
+  The <dfn method for="AppHistory">back(|options|)</dfn> method steps are:
+
+  1. If [=this=]'s [=AppHistory/current index=] is &minus;1 or 0, then return [=a promise rejected with=] an "{{InvalidStateError}}" {{DOMException}}.
+
+  1. Let |key| be [=this=]'s [=AppHistory/entry list=][[=this=]'s [=AppHistory/current index=] &minus; 1]'s [=AppHistoryEntry/session history entry=]'s [=session history entry/app history key=].
+
+  1. Return the result of [=performing an app history traversal=] given [=this=], |key|, and |options|.
+</div>
+
+<div algorithm>
+  The <dfn method for="AppHistory">forward(|options|)</dfn> method steps are:
+
+  1. If [=this=]'s [=AppHistory/current index=] is &minus;1 or is equal to [=this=]'s [=AppHistory/entry list=]'s [=list/size=] &minus; 1, then return [=a promise rejected with=] an "{{InvalidStateError}}" {{DOMException}}.
+
+  1. Let |key| be [=this=]'s [=AppHistory/entry list=][[=this=]'s [=AppHistory/current index=] + 1]'s [=AppHistoryEntry/session history entry=]'s [=session history entry/app history key=].
+
+  1. Return the result of [=performing an app history traversal=] given [=this=], |key|, and |options|.
+</div>
+
+<div algorithm>
+  To <dfn>perform an app history traversal</dfn> given an {{AppHistory}} object |appHistory|, a string |key|, and an {{AppHistoryNavigationOptions}} |options|:
+
+  1. If |appHistory|'s [=relevant global object=]'s [=associated Document=] is not [=Document/fully active=], then return [=a promise rejected with=] an "{{InvalidStateError}}" {{DOMException}}.
+
+  1. Let |jsh| be |appHistory|'s [=relevant global object=]'s [=Window/browsing context=]'s [=top-level browsing context=]'s <a spec="HTML">joint session history</a>.
+
+  1. Let |currentEntryIndex| be the index of the <a spec="HTML" lt="current entry of the joint session history">current entry</a> within |jsh|.
+
+  1. Let |destinationIndex| be the index of the closest [=session history entry=] to the <a spec="HTML" lt="current entry of the joint session history">current entry</a>, such that either that entry's [=session history entry/app history key=] equals |key|, or that entry is "standing in place of" another entry whose [=session history entry/app history key=] equals |key|. Here, |entryA| "standing in place of" <var ignore>entryB</var> means that they occupy the same "slot" within the joint session history, but only |entryA| is included because it most recently became the [=session history/current entry=] of its [=session history=].
```

Comment:
Actually, maybe the "standing in place of" thing is unnecessary, because you can only perform a navigation which navigates at least your frame (and potentially child frames)?

So maybe the algorithm could just look through your browsing context's session history for the one with the right key, and then assert that that SHE exists in the joint session history?

I'm not sure how worthwhile it is to polish this step given that the infrastructure is going to change a lot, but I couldn't stop thinking about it...

Reply:
This is wrong as shown by the example in https://github.com/WICG/app-history/issues/110#issuecomment-834831307